pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -187,40 +212,123 @@ func (c *AntreaIPAMController) cleanIPPoolForStatefulSet(namespacedName string)
+	if exists {
+		return strings.Split(annotations, annotation.AntreaIPAMAnnotationDelimiter)
+	}
+
+	return nil
+
+}
+
+// Look for an IP Pool associated with this StatefulSet, either a dedicated one or
+// annotated to the namespace. If such IP Pool is found, preallocate IPs for the StatefulSet.
+// This function returns error if pool is not found, or allocation fails.
+func (c *AntreaIPAMController) preallocateIPPoolForStatefulSet(ss *appsv1.StatefulSet) error {
+	klog.InfoS("Processing create notification", "Namespace", ss.Namespace, "StatefulSet", ss.Name)
+
+	ipPools := c.getIPPoolsForStatefulSet(ss)
+
+	if ipPools == nil {
+		// nothing to preallocate
+		return nil
+	}
+
+	if len(ipPools) > 1 {
+		return fmt.Errorf("Annotation of multiple IP Pools is not supported")
+	}
+
+	// Only one pool is supported for now. Dual stack support coming in future.
+	ipPoolName := ipPools[0]
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPoolName, c.crdClient, c.ipPoolLister)
+	if err != nil {
+		return fmt.Errorf("Failed to find IP Pool %s: %s", ipPoolName, err)
+	}
+
+	size := int(*ss.Spec.Replicas)
+	err = allocator.AllocateStatefulSet(ss.Namespace, ss.Name, size)
+	if err != nil {
+		return fmt.Errorf("Failed to preallocate continuous IP space of size %d from Pool %s: %s", size, ipPoolName, err)
+	}
+
+	return nil
+}
 func (c *AntreaIPAMController) statefulSetWorker() {
 	for c.processNextStatefulSetWorkItem() {
 	}
 }
 
 func (c *AntreaIPAMController) processNextStatefulSetWorkItem() bool {
-	key, quit := c.statefulSetCleanupQueue.Get()
+	key, quit := c.statefulSetQueue.Get()
 	if quit {
 		return false
 	}
 
-	defer c.statefulSetCleanupQueue.Done(key)
-	err := c.cleanIPPoolForStatefulSet(key.(string))
+	defer c.statefulSetQueue.Done(key)
 
-	if err != nil {
-		// Put the item back on the workqueue to handle any transient errors.
-		c.statefulSetCleanupQueue.AddRateLimited(key)
-		klog.ErrorS(err, "Failed to clean IP Pool for StatefulSet", "StatefulSet", key)
-		return true
+	namespacedName := key.(string)[1:]
+
+	if key.(string)[:1] == delEventIndication {
+		err := c.cleanIPPoolForStatefulSet(namespacedName)
+		if err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.statefulSetQueue.AddRateLimited(key)
+			klog.ErrorS(err, "Failed to clean IP Pool", "StatefulSet", key)
+			return true
+		}
+	} else {
+		ss, ok := c.statefulSetMap.Load(namespacedName)
+		if !ok {
+			// Object not found in map - should never happen
+			klog.Errorf("Failed to locate StatefulSet %s", namespacedName)
+			c.statefulSetQueue.Forget(key)
+			return true
+		}
+		err := c.preallocateIPPoolForStatefulSet(ss.(*appsv1.StatefulSet))
+		if err != nil {
+			// Preallocation is best effort - we do not retry even with transient errors,
+			// since we don't want to implement logic that would delay Pods while waiting for
+			// preallocation.
+			klog.ErrorS(err, "No IPs preallocated")
+			c.statefulSetMap.Delete(key)
```

Comment:
We can delete a recreated ss?

Reply:
Do you mean create + delete + create?

Reply:
Good point about retry during delete event. I removed the retry and added comment that we should rely on garbage collector instead of retry.